crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -809,32 +804,26 @@ impl Object {
         ctx: &Context<'_>,
         keys: Vec<ObjectKey>,
         checkpoint_viewed_at: u64,
-    ) -> Result<Vec<Self>, Error> {
+    ) -> Result<Vec<Option<Self>>, Error> {
         let DataLoader(loader) = &ctx.data_unchecked();
 
-        let keys: Vec<PointLookupKey> = keys
+        let keys: Vec<_> = keys
             .into_iter()
             .map(|key| PointLookupKey {
                 id: key.object_id,
                 version: key.version.into(),
             })
             .collect();
 
-        let data = loader.load_many(keys).await?;
-        let objects: Vec<_> = data
+        let data = loader.load_many(keys.clone()).await?;
+        Ok(keys
             .into_iter()
-            .filter_map(|(lookup_key, bcs)| {
-                Object::new_serialized(
-                    lookup_key.id,
-                    lookup_key.version,
-                    bcs,
-                    checkpoint_viewed_at,
-                    lookup_key.version,
-                )
+            .flat_map(|k| {
+                data.get(&k).cloned().map(|bcs| {
+                    Object::new_serialized(k.id, k.version, bcs, checkpoint_viewed_at, k.version)
+                })
             })
-            .collect();
-
-        Ok(objects)
+            .collect())
```

Comment:
This should not use `flat_map`, it should be `map` instead and if the data loader does not return a result for a given key, we pass that result on by returning `None` here for that position in the vector too.

Let's also add a transactional test for this case.